### PR TITLE
Image based lighting

### PR DESCRIPTION
**Pull request by miquelmiro3**
Image based lighting is working, but there is a problem with specular colors that need to be fixed to obtain good results. To test it in the Configuration window there is a toggle box to activate IBL but first click the Build button.

### CODE REVIEW

**Comment by miquelmiro3** [comment]
Trying to resolve the specular bug before merging. For those who don't know. I saw that when selecting a red specular color, the resultant one was blue/cian and i think this proble is effecting IBL and that is why i was sawing strange results.

File: Source/src/core/rendering/Skybox.cpp [hunk 1]
```diff
@@ -216,3 +274,283 @@ void Hachiko::Skybox::ChangeCubeMapSide(UID texture_uid, TextureCube::Side cube_
...
+    if (!cube.loaded)
+    {
+        HE_ERROR("There is not a skybox loaded");
+        return;
+    }
+    unsigned width = cube.resources[0]->width;
+    unsigned height = cube.resources[0]->height;
+
+    // Delete last irradiance cubemap
+    glDeleteTextures(1, &prefilteredIBL_id);
```
Comment (baransrc) [inline Source/src/core/rendering/Skybox.cpp:396]: Let's rename this to `prefiltered_ibl_id` to adhere to our naming conventions.

File: Source/src/core/rendering/Skybox.cpp [hunk 2]
```diff
@@ -216,3 +274,283 @@ void Hachiko::Skybox::ChangeCubeMapSide(UID texture_uid, TextureCube::Side cube_
...
+            GL_RGB32F,
+            cube.resources[i]->width,
+            cube.resources[i]->height,
+            0,
+            GL_RGB, 
+            GL_UNSIGNED_BYTE,
+            0
+        );
+    }
+    prefilteredIBL_numLevels = int(log(float(cube.resources[0]->width)) / log(2));
```
Comment (baransrc) [inline Source/src/core/rendering/Skybox.cpp:434]: Let's rename this to `prefiltered_ibl_number_of_levels`

File: Source/src/core/rendering/Skybox.h [hunk 3]
```diff
@@ -22,16 +22,30 @@ namespace Hachiko
...
         void CreateBuffers();
         TextureCube cube;
         unsigned vao{};
         unsigned vbo{};
 
+        bool activate_IBL = false;
+        unsigned diffuseIBL_id = 0;
+        unsigned prefilteredIBL_id = 0;
+        unsigned environmentBRDF_id = 0;
+        unsigned prefilteredIBL_numLevels = 0;
```
Comment (baransrc) [inline Source/src/core/rendering/Skybox.h:47]: These should be renamed like:
```suggestion
        bool activate_ibl = false;
        unsigned diffuse_ibl_id = 0;
        unsigned prefiltered_ibl_id = 0;
        unsigned environment_brdf_id = 0;
        unsigned prefiltered_ibl_number_of_levels = 0;
```

File: Source/src/modules/ModuleProgram.cpp [hunk 4]
```diff
@@ -281,6 +313,17 @@ void Hachiko::ModuleProgram::UpdateCamera(const ComponentCamera* camera) const
     UpdateUBO(UBOPoints::CAMERA, sizeof(CameraData), &camera_data);
 }
 
+void Hachiko::ModuleProgram::UpdateCamera(const Frustum& frustum) const
+{
+    CameraData camera_data;
+    camera_data.view = float4x4(frustum.ViewMatrix());
+    camera_data.proj = frustum.ProjectionMatrix();
+    // TODO: Understand why camera_data.view.TranslatePart() does not give the position
+    camera_data.pos = camera_data.view.RotatePart().Transposed().Transform(-camera_data.view.TranslatePart());
```
Comment (baransrc) [inline Source/src/modules/ModuleProgram.cpp:322]: We can directly get the correct position from the component camera probably.

File: Source/src/modules/ModuleProgram.h [hunk 5]
```diff
@@ -129,6 +148,9 @@ namespace Hachiko
         Program* deferred_geometry_program = nullptr;
         Program* deferred_lighting_program = nullptr;
         Program* skybox_program = nullptr;
+        Program* diffuseIBL_program = nullptr;
+        Program* prefilteredIBL_program = nullptr;
+        Program* environmentBRDF_program = nullptr;
```
Comment (baransrc) [inline Source/src/modules/ModuleProgram.h:153]: These should be renamed like:
```suggestion
        Program* diffuse_ibl_program = nullptr;
        Program* prefiltered_ibl_program = nullptr;
        Program* environment_brdf_program = nullptr;
```

File: Source/src/modules/ModuleRender.cpp [hunk 6]
```diff
@@ -147,6 +148,11 @@ UpdateStatus Hachiko::ModuleRender::PreUpdate(const float delta)
 {
     render_list.PreUpdate();
 
+    //Skybox* skybox = App->scene_manager->GetActiveScene()->GetSkybox();
+    //skybox->GenerateDiffuseIBL();
+    //skybox->GeneratePrefilteredIBL();
+    //skybox->GenerateEnvironmentBRDF();
+
```
Comment (baransrc) [inline Source/src/modules/ModuleRender.cpp:155]: Let's delete these comments.

File: Game/shaders/fragment_environmentBRDF.glsl [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+#define NUM_SAMPLES 100
+#define PI 3.14159
+
+in vec2 tex_coords;
+
+out vec4 fragColor;
+
+float radicalInverse_VdC(uint bits) 
+{
+    bits = (bits << 16u) | (bits >> 16u);
```
[the remark this answers is not in the document]
Comment (miquelmiro3) [reply Game/shaders/fragment_environmentBRDF.glsl]: It is

File: Source/src/modules/ModuleProgram.cpp (hunk 4 above)
Comment (miquelmiro3) [reply Source/src/modules/ModuleProgram.cpp:322]: In that case we don't have a component camera, because we use a specific frustrum created previously

**Review by baransrc — APPROVED** [review]
Thanks for the changes, everything looks good to me 💯

File: Game/shaders/common/functionality/lighting_functions.glsl [hunk 8]
```diff
@@ -113,4 +112,20 @@ vec3 SpotPBR(const vec3 frag_pos, const vec3 normal, const vec3 view_dir, const
...
 }
 
+// TODO: pass uniforms as in parameters
+vec3 GetAmbientLight(in vec3 normal, in vec3 R, float NdotV, float roughness, in vec3 diffuse_color, in vec3 specular_color)
+{
+    if (activate_IBL > 0)
+    {
+        vec3 irradiance = texture(diffuseIBL, normal).rgb;
+        vec3 radiance = textureLod( prefilteredIBL, R, roughness * prefilteredIBL_numLevels).rgb;
+        vec2 fab = texture( environmentBRDF, vec2(NdotV, roughness)).rg;
```
Comment (ermario) [inline Game/shaders/common/functionality/lighting_functions.glsl:122]: this should be .rg or .rgb ?

File: Game/shaders/fragment_diffuseIBL.glsl [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    vec3 normal = normalize(tex_coords);
+    mat3 tangentSpace = computeTangetSpace(normal);
+    for(int i=0; i< NUM_SAMPLES; ++i)
+    {
+        vec2 rand_value = hammersley2D(i, NUM_SAMPLES);
+        vec3 sample_dir = tangentSpace * hemisphereSample(rand_value[0], rand_value[1]); 
+        irradiance += texture(skybox, sample_dir).rgb;
+    }
+    fragColor = vec4(irradiance*(1.0/float(NUM_SAMPLES)), 1.0);
+}
```
Comment (ermario) [inline Game/shaders/fragment_diffuseIBL.glsl:25]: Warning here "No newline at end of file"

File: Game/shaders/vertex_environmentBRDF.glsl [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+out vec2 tex_coords;
+
+void main()
+{
+    float x = -1.0 + float((gl_VertexID & 1) << 2);
+    float y = -1.0 + float((gl_VertexID & 2) << 1);
+    tex_coords.x = (x + 1.0) * 0.5;
+    tex_coords.y = (y + 1.0) * 0.5;
+    gl_Position = vec4(x, y, 0.0, 1.0);
+}
```
Comment (ermario) [inline Game/shaders/vertex_environmentBRDF.glsl:12]: maybe another warning? idk why github marks me that :S

**Review by ermario — APPROVED** [review]
I added a few comments, most of them are just little stuff so it's all good. Couldn't understand really the code since is a big feature but looks good overall.

File: Game/shaders/common/functionality/lighting_functions.glsl (hunk 8 above)
Comment (miquelmiro3) [reply Game/shaders/common/functionality/lighting_functions.glsl:122]: rg

**Review by VBoschMatas — APPROVED** [review]

**Pull request merged**